## Introduction to NumPy

NumPy were written in C. NumPy operations on arrays and lists are a lot faster than performing the same operations in Python. Avoiding loops.

In [2]:
import numpy as np

 # 1. Data types and Attributes

In [52]:
a1 = np.array([1, 2, 3])
a1

array([1, 2, 3])

In [53]:
type(a1)  #ndarray is main datatype in NumPy

numpy.ndarray

In [54]:
a2 = np.array([[1, 2.0, 3.3],
              [4, 5, 6.5]])
a2

array([[1. , 2. , 3.3],
       [4. , 5. , 6.5]])

In [55]:
a3 = np.array([[[1, 2, 3],
                [4, 5, 6]],
               [[7, 8, 9],
                [10, 11, 12]],
               [[13, 14, 15],
                [16, 17, 18]]])
a3

array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18]]])

In [56]:
a3.shape

(3, 2, 3)

In [31]:
a3.ndim  # number of dimensions

3

In [36]:
# Create a FataFrame from a NumPy array
import pandas as pd

df = pd.DataFrame(a2)
df

,0,1,2
0,1.0,2.0,3.3
1,4.0,5.0,6.5


## 2. Creating a numPy arrays

In [38]:
sample_array = np.array([1, 2, 3])
sample_array

array([1, 2, 3])

In [39]:
sample_array.dtype

dtype('int64')

In [46]:
ones = np.ones((2, 3), dtype='int64', order='C')    # shift + tab  ---> gives an instruction how to use function
ones

array([[1, 1, 1],
       [1, 1, 1]])

In [48]:
zeroes = np.zeros((2, 3), dtype='int64', order='C')
zeroes

array([[0, 0, 0],
       [0, 0, 0]])

In [51]:
# range array 
range_array = np.arange(1, 25, 3)
range_array

array([ 1,  4,  7, 10, 13, 16, 19, 22])

In [5]:
# Random array - wypisuje randomowe inty od 1 do 20 (5 liczb)
random_array = np.random.randint(1, 20, 5)
random_array

# Macierz 
random_arr = np.random.randint(1, 20, size=(10,15))
random_arr


array([[12, 12, 17, 10, 16, 15, 15, 19, 12,  3,  5, 19,  7,  9,  7],
       [18,  4, 14, 18,  9,  2, 15,  7, 12,  8, 15,  3, 14, 17,  4],
       [18,  8,  4,  2,  6, 10,  4, 18, 12,  2, 10,  4, 14, 16, 15],
       [ 8, 14,  8, 16, 13, 18, 15, 13,  9, 15, 13,  1,  7,  9,  1],
       [12,  8, 11, 19, 17,  8,  3,  3,  1,  5, 10,  7,  9,  7,  9],
       [ 8, 12,  2,  1, 16,  5,  3, 12,  8,  3,  1,  3,  5, 15, 14],
       [ 3,  1,  5, 14,  7,  9, 15, 15, 10, 13, 19,  7, 17,  4,  5],
       [ 7, 13, 15, 11,  4, 13,  7, 19,  2, 10, 13,  6, 12, 12, 11],
       [ 7,  1,  1, 13,  9,  3,  7,  6,  8,  9,  5,  1, 19, 10, 12],
       [15,  9, 17, 17, 12,  7,  2,  3, 17,  5, 17, 17, 17,  2,  2]])

In [70]:
# Rozmiar randomowej macierzy 
random_arr.size

150

In [74]:
# Returning random floats 
np.random.random((5,5))

array([[0.94465125, 0.92610991, 0.57510655, 0.98175366, 0.99683189],
       [0.04114602, 0.11861453, 0.51157209, 0.36553551, 0.47622466],
       [0.06211206, 0.1646395 , 0.45357419, 0.27277453, 0.03640252],
       [0.36570534, 0.61330788, 0.70776766, 0.25470511, 0.51219027],
       [0.98576135, 0.32721271, 0.04311954, 0.55334036, 0.01605794]])

## Random Seed

In [3]:
# Pseudo random numbers - still generate random numbers, but random niumbers that are reproducible
np.random.seed(seed=42)   #jak zakomentujesz i zabierzesz seep w nawiasach to liczby będa inne 
randomSeed = np.random.randint(10, size=(5,3))
randomSeed

array([[6, 3, 7],
       [4, 6, 9],
       [2, 6, 7],
       [4, 3, 7],
       [7, 2, 5]])

## 3. Viewing arrays and matrices


In [6]:
np.unique(random_arr)

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19])

In [14]:
randomSeed[0][1]


np.int64(3)

In [16]:
rozmiarek = randomSeed.shape

rozmiarek

(5, 3)

In [18]:
randomSeed

array([[6, 3, 7],
       [4, 6, 9],
       [2, 6, 7],
       [4, 3, 7],
       [7, 2, 5]])

In [34]:
# slicing:  2-dimensional array
randomSeed[:3, :2]

array([[6, 3],
       [4, 6],
       [2, 6]])

In [38]:
# 4-dimensional array, max value is 19 
a4 = np.random.randint(20, size=(2, 5, 3, 3))
a4

array([[[[16, 19,  9],
         [ 5, 14, 10],
         [ 4,  0,  7]],

        [[11, 11,  4],
         [ 6,  3,  5],
         [12, 19, 14]],

        [[ 2,  7, 19],
         [15, 12, 17],
         [ 9, 18, 16]],

        [[18,  4,  8],
         [11,  0,  0],
         [14,  1, 15]],

        [[ 7, 12,  0],
         [15,  6,  4],
         [ 2, 11, 15]]],


       [[[18,  4, 13],
         [ 4, 14, 16],
         [13, 19,  4]],

        [[11, 15, 15],
         [ 6,  3,  0],
         [ 4,  9,  4]],

        [[ 3,  1, 19],
         [ 9, 18,  0],
         [ 4, 12,  3]],

        [[15, 15,  1],
         [16, 19, 11],
         [17,  2,  0]],

        [[ 0, 18, 10],
         [ 4, 11,  2],
         [ 0,  0,  7]]]])

In [42]:
a4.ndim

4

In [49]:
# getting the first number of the inner most arrays 
a4[:, :, 2:, :1]

array([[[[ 4]],

        [[12]],

        [[ 9]],

        [[14]],

        [[ 2]]],


       [[[13]],

        [[ 4]],

        [[ 4]],

        [[17]],

        [[ 0]]]])

## 4. Manipulating and comparing arrays

# Arithmetic

In [57]:
a1

array([1, 2, 3])

In [64]:
shape = (3, 3) 
ones = np.ones(shape)
ones

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [66]:
suma = a1+ones
suma

array([[2., 3., 4.],
       [2., 3., 4.],
       [2., 3., 4.]])

In [68]:
a1

array([1, 2, 3])

In [69]:
a2

array([[1. , 2. , 3.3],
       [4. , 5. , 6.5]])

In [70]:
multipl = a1*a2
multipl

array([[ 1. ,  4. ,  9.9],
       [ 4. , 10. , 19.5]])

In [75]:
#ValueError: operands could not be broadcast together with shapes (2,3) (2,5,3,3)   a2 * a4
# Broadcasting describe show how numpy treats arrays with different shapes during arithmetic operations

In [76]:
a2 * a4  # broadcast problem 

ValueError: operands could not be broadcast together with shapes (2,3) (2,5,3,3) 

In [78]:
a1/ones 

array([[1., 2., 3.],
       [1., 2., 3.],
       [1., 2., 3.]])

In [84]:
# Floor division - rounds down (zaokrgla w dół)
a2//a1

array([[1., 1., 1.],
       [4., 2., 2.]])

In [85]:
# Power of two ^2
a2**2 

array([[ 1.  ,  4.  , 10.89],
       [16.  , 25.  , 42.25]])

In [89]:
# Power of two ^2
np.square(a2)

array([[ 1.  ,  4.  , 10.89],
       [16.  , 25.  , 42.25]])

In [90]:
np.add(a1, a2) # adding numbers

array([[2. , 4. , 6.3],
       [5. , 7. , 9.5]])

In [92]:
# MODULO
a2 % 2

array([[1. , 0. , 1.3],
       [0. , 1. , 0.5]])

In [94]:
# MODULO
np.mod(a2, 2)

array([[1. , 0. , 1.3],
       [0. , 1. , 0.5]])

## Aggregation

### Aggregation = performing the same operation on a number of things

In [95]:
listy_list = [1, 2, 3]
type(listy_list)

list

In [96]:
sum(listy_list)

6

In [110]:
masiveArray = np.random.random(10000)
print(f" 10 numers: {masiveArray[:10]}")

 10 numers: [0.35781125 0.59378063 0.57559343 0.16145783 0.47173984 0.55347178
 0.57028694 0.21010719 0.74227139 0.02519909]


In [112]:
# Time %timeit
%timeit sum(masiveArray)
%timeit np.sum(masiveArray)


486 μs ± 36.4 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
2.9 μs ± 307 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [109]:
# We should use numPy methods for numPy data
%timeit np.sum(masiveArray)

1.33 μs ± 311 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [126]:
print(f"{a2}")
print(f"Variance: {np.var(a2)}")  # Wariacja
print(f"Standard deviation: {np.std(a2)}") # Odchylenie standardowe
print()
print(f"The standard deviation is actually a square root of the variance")
print(f":{np.sqrt}")

[[1.  2.  3.3]
 [4.  5.  6.5]]
Variance: 3.3222222222222224
Standard deviation: 1.8226964152656422

The standard deviation is actually a square root of the variance


## Wariancja (variance)

Miara statystyczna, która określa, jak bardzo dane są rozproszone (rozrzucone) wokół ich średniej arytmetycznej. Im wyższa wariancja, tym bardziej wartości odstają od średniej. 
Wariancja jest nam potrzebna, aby zrozumieć stopień niepewności, szumu lub zmienności w analizowanym systemie.

1. Wykrywanie "szumu" i stabilności sieci (Telekomunikacja)
Wyobraź sobie, że analizujesz opóźnienia (ping) na dwóch różnych serwerach. Oba mają średni ping na poziomie 50 ms.

        Serwer A: 49 ms, 50 ms, 51 ms (Wariancja bliska zeru — połączenie jest niezwykle stabilne).

        Serwer B: 10 ms, 50 ms, 90 ms (Bardzo wysoka wariancja — połączenie szarpie).

## Odchylenie standardowe (standard deviation)

Najbardziej intuicyjna i najczęściej stosowana miara tego, jak bardzo dane są rozproszone wokół średniej. Dzięki temu odchylenie mówi Ci wprost: "Twoje dane odchylają się od średniej przeciętnie o X jednostek".

W analizie danych i uczeniu maszynowym bardzo często zakłada się, że dane mają tzw. rozkład normalny (wykres przypominający dzwon). Dla takich danych odchylenie standardowe pozwala natychmiast ocenić, co jest "normą", a co "anomalią":

    ~68% wszystkich Twoich danych znajduje się w odległości 1 odchylenia standardowego od średniej.

    ~95% danych mieści się w odległości 2 odchyleń.

    ~99.7% danych mieści się w odległości 3 odchyleń (wszystko poza tym to bardzo rzadkie anomalie, tzw. outliery).

Przykład: Jeśli średni wzrost w grupie to 170 cm, a odchylenie standardowe to 10 cm, to wiesz, że około 68% osób z tej grupy ma wzrost między 160 cm a 180 cm.